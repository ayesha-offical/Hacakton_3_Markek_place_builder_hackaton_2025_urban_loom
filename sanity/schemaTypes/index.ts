import { type SchemaTypeDefinition } from 'sanity'

import {categoryType} from './categoryType'

import { salesService } from './sale'
import { productType } from './ProductType'

export const schema: { types: SchemaTypeDefinition[] } = {
  types: [ categoryType, salesService,productType],
}