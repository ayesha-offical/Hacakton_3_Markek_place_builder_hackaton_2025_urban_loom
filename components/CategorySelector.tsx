"use client";
import { Category } from "@/sanity.types";
import { useRouter } from "next/navigation";
import React, { useState } from "react";
import { Popover, PopoverContent, PopoverTrigger } from "@/components/ui/popover";
import { Button } from "@/components/ui/button";
import { Check, ChevronsUpDown } from "lucide-react";
import {
  Command,
  CommandInput,
  CommandList,
  CommandItem,
  CommandEmpty,
  CommandGroup,
} from "@/components/ui/command";
import { cn } from "@/lib/utils";
interface Props {
  categories: Category[]; // Array of category names
}

const CategorySelector = ({ categories }: Props) => {
  const [open, setOpen] = useState(false);
  const [value, setValue] = useState("");
  const router = useRouter();

  return (
    <Popover open={open} onOpenChange={setOpen}>
      <PopoverTrigger asChild>
        <Button
          variant="outline"
          role="combobox"
          aria-expanded={open}
          className="w-[200px] justify-between"
        >
          {value
            ? categories.find((category) => category?._id === value)?.title
            : "Filter by category"}
          <ChevronsUpDown />
        </Button>
      </PopoverTrigger>
      <PopoverContent className="w-[200px] p-0">
        <Command>
          <CommandInput
            placeholder="Search By Category......."
            className="h-9"
          onKeyDown={(e: any) => {
              if (e.key === "Enter") {
                const selectedCategory = categories.find((c) =>
                  c.title
                    ?.toLowerCase()
                    .includes(e.currentTarget.value.toLowerCase())
                );
                if (selectedCategory?.slug?.current){
                  setValue(selectedCategory?._id);
                  router.push(`/categories/${selectedCategory.slug.current}`);
                  setOpen(false);
              
                }
                  
              }
         }}

          />

          <CommandList>
            <CommandEmpty className="text-red-700 flex justify-center items-center py-3 ">
              No Category Found.
            </CommandEmpty>
            <CommandGroup>
              {categories?.map((category) => (
                <CommandItem
                  key={category?._id}
                  value={category?.title}
                  onSelect={() => {
                    setValue(value === category?._id ? category?._id : "");
                    router.push(`/categories/${category.slug?.current}`);
                    setOpen(false);
                  }}
                  // onClick={() => {
                  //   setValue(category?._id);
                  //   router.push(`/products?category=${category?._id}`);
                  //   setOpen(false);
                  // }}
                >
                  {category?._id === value ? (
                    <strong>{category.title}</strong>
                  ) : (
                    category.title
                  )}

                </CommandItem>
              ))}
            </CommandGroup>
          </CommandList>
        </Command>
      </PopoverContent>
    </Popover>
  );
};

export default CategorySelector;
