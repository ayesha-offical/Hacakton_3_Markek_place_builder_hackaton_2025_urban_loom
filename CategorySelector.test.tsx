import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";

window.HTMLElement.prototype.scrollIntoView = function() {};

// Mock ResizeObserver
global.ResizeObserver = class {
  observe() {}
  unobserve() {}
  disconnect() {}
};

import CategorySelector from "@/components/CategorySelector"; // Adjust the path if needed
import { Category } from "@/sanity.types";
import "@testing-library/jest-dom";


// Create a mock for next/navigation's useRouter with a push function
const pushMock = jest.fn();
jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: pushMock,
  }),
}));

describe("CategorySelector Component", () => {
  // Define some sample categories
  const mockCategories: Category[] = [
    { _id: "1", title: "Electronics", slug: {  _type: 'slug', current: "electronics" } },
    { _id: "2", title: "Books", slug: { _type: 'slug', current: "books" } },
    { _id: "3", title: "Clothing", slug: { _type: 'slug', current: "clothing" } },
  ];

  beforeEach(() => {
    pushMock.mockClear();
  });

  test("renders the default button text when no category is selected", () => {
    const { container } = render(<CategorySelector categories={mockCategories} />);
    // Find the element that contains the default text and then get its closest button
    const textElement = screen.getByText(/filter by category/i);
    const button = textElement.closest("button");
    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent(/filter by category/i);
    expect(container).toMatchSnapshot();
  });

  test("updates selection when a category is selected via click", async () => {
    render(<CategorySelector categories={mockCategories} />);
    // Open the popover by clicking the button
    const textElement = screen.getByText(/filter by category/i);
    const button = textElement.closest("button");
    expect(button).toBeInTheDocument();
    fireEvent.click(button!);
    
    // Wait for the CommandItem to appear and simulate a click on "Books"
    const commandItem = await screen.findByText("Books");
    fireEvent.click(commandItem);
    
    // Expect router.push to be called with the correct URL for Books
    await waitFor(() => {
      expect(pushMock).toHaveBeenCalledWith("/categories/books");
    });
  });

  test("searches for a category and routes when Enter is pressed", async () => {
    render(<CategorySelector categories={mockCategories} />);
    // Open the popover
    const textElement = screen.getByText(/filter by category/i);
    const button = textElement.closest("button");
    expect(button).toBeInTheDocument();
    fireEvent.click(button!);
    
    // Find the search input (using its placeholder text)
    const input = screen.getByPlaceholderText("Search By Category.......");
    // Simulate typing and then pressing Enter to search for "Electronics"
    fireEvent.change(input, { target: { value: "elect" } });
    fireEvent.keyDown(input, { key: "Enter", code: "Enter" });
    
    await waitFor(() => {
      expect(pushMock).toHaveBeenCalledWith("/categories/electronics");
    });
  });

  test("matches snapshot after opening the popover", async () => {
    const { container } = render(<CategorySelector categories={mockCategories} />);
    const textElement = screen.getByText(/filter by category/i);
    const button = textElement.closest("button");
    expect(button).toBeInTheDocument();
    fireEvent.click(button!);
    // Wait for the popover content (search input) to appear
    await waitFor(() => {
      expect(screen.getByPlaceholderText("Search By Category.......")).toBeInTheDocument();
    });
    expect(container).toMatchSnapshot();
  });
});
