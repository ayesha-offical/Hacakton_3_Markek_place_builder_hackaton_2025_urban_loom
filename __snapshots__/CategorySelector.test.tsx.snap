// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategorySelector Component matches snapshot after opening the popover 1`] = `
<div>
  <button
    aria-controls="radix-:rj:"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="inline-flex items-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-[200px] justify-between"
    data-state="open"
    role="Button"
    type="button"
  >
    Filter by category
    <svg
      class="lucide lucide-chevrons-up-down"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m7 15 5 5 5-5"
      />
      <path
        d="m7 9 5-5 5 5"
      />
    </svg>
  </button>
</div>
`;

exports[`CategorySelector Component renders the default button text when no category is selected 1`] = `
<div>
  <button
    aria-controls="radix-:r0:"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="inline-flex items-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-[200px] justify-between"
    data-state="closed"
    role="Button"
    type="button"
  >
    Filter by category
    <svg
      class="lucide lucide-chevrons-up-down"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m7 15 5 5 5-5"
      />
      <path
        d="m7 9 5-5 5 5"
      />
    </svg>
  </button>
</div>
`;
